core/types/calls.cc

```diff
@@ -1852,6 +1852,40 @@ class T_must : public IntrinsicMethod {
     }
 } T_must;
 
+// T.must_because has handling equivalent to T.must for its first argument
+class T_must_because : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        if (args.args.empty()) {
+            return;
+        }
+        if (!args.args[0]->type.isFullyDefined()) {
+            if (auto e = gs.beginError(args.argLoc(0), errors::Infer::BareTypeUsage)) {
+                e.setHeader("`{}` applied to incomplete type `{}`", "T.must_because", args.args[0]->type.show(gs));
+            }
+            return;
+        }
+        auto ret = Types::dropNil(gs, args.args[0]->type);
+        if (ret == args.args[0]->type) {
+            if (auto e = gs.beginError(args.argLoc(0), errors::Infer::InvalidCast)) {
+                if (args.args[0]->type.isUntyped()) {
+                    e.setHeader("`{}` called on `{}`, which is redundant", "T.must_because", args.args[0]->type.show(gs));
+                } else {
+                    e.setHeader("`{}` called on `{}`, which is never `{}`", "T.must_because", args.args[0]->type.show(gs),
+                                "nil");
+                }
+                e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
+                auto replaceLoc = args.callLoc();
+                const auto locWithoutTMustBecause = args.callLoc().adjust(gs, 7, -1);
```

Comment:
@jez What I'm assuming based on the numbers is:

```diff
- T.must(x)
+ x
```

But with `T.must_because(x) {...}` what could I do to properly rewrite this? Or should I axe this auto-correct for ease?

Reply:
I have no idea why the original code was so complicated, and I probably wrote it.

I'm pretty sure you should be able to do

```cpp
e.replaceWith("Remove `T.must_because`", replaceLoc, "{}", args.locs[0].source(gs).value());
```

(you might have to futz with it to get it to typecheck, not sure. but the idea is that `args` has something inside it that tells you what the location of the first argument to the method call is.

Reply:
Looks like it might be called `args.argLoc(0)` now